### DISCUSSION TITLE
route middleware has different behavior in opposite to nuxt2 ?

### DISCUSSION THREAD

**Discussion by mercs600** [nuxt/nuxt#17118] [Questions]
✋
I would like to ask about behavior of route middlewares in nuxt3 to opposite to nuxt2. 
In the nuxt2 doc we can read 

`In universal mode, middlewares will be called once on server-side (on the first request to the Nuxt app, e.g. when directly accessing the app or refreshing the page) and on the client-side when navigating to further routes. With ssr: false, middlewares will be called on the client-side in both situations.`

So middleware is called on server side (on the request) once and after that only on route change. 

But what I see in the nuxt3 that route middleware: 
1. is called once on server side - on the request (what is ok)
2. is called on render the page immediately - on client side (what is not ok according to nuxt2) 
3. is called on each route change - on client side (what is ok)
4. is called also for other requests - not pages only - like favicon etc. (is it ok ?) 

My questions are:
1. Is it intended behavior ? 
2. Is there some possibility to call middleware in the same way like in the nuxt2 ? 

~~Now I have to do some weird conditionals to achieve this:~~ 
```js
  if ((process.server && to) || (!process.server && to.fullPath !== currentPageSlug)) {
    // do something
  }
```
~~It works but I have to rely on some data like currentPageslug to avoid calling middleware after render the page on client side.~~
**EDIT:**
This condition above could be replaced with 
```js
 if (process.server || !isEqual(to.fullPath, from.fullPath)) {} 
```

Do you have better idea ?

**EDIT:**
It could be fixed here:
https://github.com/nuxt/framework/blob/main/packages/nuxt/src/pages/runtime/router.ts#L145

```js
  let result
  if (process.server || !isEqual(to.fullPath, from.fullPath)) {
    result = await callWithNuxt(nuxtApp, middleware, [to, from])
  }
```

In this case is called only on server side and also only for route change, like in the nuxt2.

**Comment by psycongaroo** [comment]
That will be so usefull, like check session just on server side and not duplicate the request in client side too.
Hope staffs can give it a look.

@danielroe
